### PR TITLE
Removed dependency to `Microsoft.Dotnet.Cli.Utils`

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
- fixes https://github.com/dotnet/templating/issues/5296

### Solution
Created separated `CommandUtils` package usable from CLI and XUnit tests.
Removed dependency to `Microsoft.Dotnet.Cli.Utils`.

### Checks:
- [ ] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: test/Microsoft.TemplateEngine.CommandUtils/ProcessReaper.cs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    /// On Windows, a job object will be used to ensure the termination of the target
+    /// process (and its tree) even if the current process is rudely terminated.
+    ///
+    /// On POSIX systems, the reaper will handle SIGTERM and attempt to forward the
+    /// signal to the target process only.
+    ///
+    /// The reaper also suppresses SIGINT in the current process to allow the target
+    /// process to handle the signal.
+    /// </remarks>
+    internal class ProcessReaper : IDisposable
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.CommandUtils/ProcessReaper.cs:25]: let's discuss if we need this feature for our purpose

**Comment by vlada-shubina** [comment]
@JanKrivanek I excluded the tools from source-build in the last commit. If they are needed, we can turn it on again.

**Comment by vlada-shubina** [comment]
we agreed not to create a public package at the time, and include the source code internally instead.

**Comment by vlada-shubina** [comment]
@JanKrivanek as agreed - moved the utils to shared internal code. 
It might get too difficult to review - let me know if you'd like to split it into 2 PRs.

File: tools/Shared/Microsoft.TemplateEngine.CommandUtils/NativeMethods.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            [DllImport("kernel32.dll", SetLastError = true)]
+            internal static extern bool AssignProcessToJobObject(IntPtr hJob, IntPtr hProcess);
+
+            [DllImport("kernel32.dll", CharSet = CharSet.Auto)]
+            [DefaultDllImportSearchPaths(DllImportSearchPath.System32)]
+            internal static extern IntPtr GetCommandLine();
+
+            [DllImport("ntdll.dll", SetLastError = true)]
+            [DefaultDllImportSearchPaths(DllImportSearchPath.System32)]
+            internal static extern unsafe uint NtQueryInformationProcess(SafeProcessHandle ProcessHandle, int ProcessInformationClass, void* ProcessInformation, uint ProcessInformationLength, out uint ReturnLength);
```
Comment (JanKrivanek) [inline tools/Shared/Microsoft.TemplateEngine.CommandUtils/NativeMethods.cs:41]: This seems to be the only reason why we need `AllowUnsafeBlocks` in all importing projects, while it's not being used.
I'd probably get rid of it and remova all `AllowUnsafeBlocks`

File: tools/Shared/Microsoft.TemplateEngine.CommandUtils/TestCommand.cs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+using Microsoft.Extensions.Logging;
+using Xunit.Abstractions;
+
+namespace Microsoft.TemplateEngine.CommandUtils
+{
+    internal abstract class TestCommand
+    {
+        private readonly LoggerWrapper _loggerWrapper;
+
+        protected TestCommand(ITestOutputHelper log)
```
Comment (JanKrivanek) [inline tools/Shared/Microsoft.TemplateEngine.CommandUtils/TestCommand.cs:14]: I have a bit mixed feelings about this.
It's nice to have the overload for XUnit logger, on the other hand the reference to XUnit is then forced to all callers.

Maybe we can offer overload accepting custom interface (extracted from LoggerWrapper) and callers would be able to pass that, plus there can be an extra extension class (not required to be imported) conveniently converting from XUnit logger to the interface.

It's not a strong opinion on my side thought.

File: tools/Shared/Microsoft.TemplateEngine.CommandUtils/DotnetCommand.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+namespace Microsoft.TemplateEngine.CommandUtils
+{
+    internal class DotnetCommand : TestCommand
+    {
+        internal DotnetCommand(ILogger log, string subcommand, params string[] args) : base(log)
+        {
+            Arguments.Add(subcommand);
+            Arguments.AddRange(args);
+            //opt out from telemetry
+            WithEnvironmentVariable("DOTNET_CLI_TELEMETRY_OPTOUT", "true");
```
Comment (JanKrivanek) [inline tools/Shared/Microsoft.TemplateEngine.CommandUtils/DotnetCommand.cs:16]: We definitely want to exclude telemetry for our CIs. I'm not sure how we want to look on the tests executed by 3rd parties via `TemplateVerifier` in their environments... But it's probably best to exclude everything - as then we'd need to allow distinguishing between 1st party and 3rd party usages of `TemplateVerifier`... and test invocations are not actual instantiations anyways.

@baronfel as FYI - there is a slight (and hopefully unprobable) effect on our telemetry stats - in case there would be  3rd parties that are executing some custom tests today (or even 1st parties that do not properly exclude telemetry) and would switch to `TemplateVerifier` - than we might see a declines in stats that are not real.

**Review by JanKrivanek — APPROVED** [review]
Overall looks good.
I have 2 minor things for considerations - but not strongly opinionated about those.

I'm very much looking forward for this to be merged!! :-) As it will simplify my life in unblocking dogfooding in sdk (need to be able to inject custom environemnt and dotnet assembly)

File: tools/Shared/Microsoft.TemplateEngine.CommandUtils/TestCommand.cs (hunk 3 above)
Comment (vlada-shubina) [reply tools/Shared/Microsoft.TemplateEngine.CommandUtils/TestCommand.cs:14]: The current use cases are `XUnit` and `ILogger`. If we expose it, l agree, but for now, for simplicity I would prefer to have it as.

File: tools/Shared/Microsoft.TemplateEngine.CommandUtils/DotnetCommand.cs (hunk 4 above)
Comment (vlada-shubina) [reply tools/Shared/Microsoft.TemplateEngine.CommandUtils/DotnetCommand.cs:16]: Maybe it would be better to let the caller to decide on this?
But then we need to extra work in all the tests.

File: tools/Shared/Microsoft.TemplateEngine.CommandUtils/TestCommand.cs (hunk 3 above)
Comment (JanKrivanek) [reply tools/Shared/Microsoft.TemplateEngine.CommandUtils/TestCommand.cs:14]: Thinking about it - we anyway already have a transitive dependency on `XUnit.Abstractions` due to inability to dirrectly call `Verify` (design decision - all functionality is internal and exposed via `internalsvisibleto` only to protocol specific libraries - so we chose to reference `Verify.XUnit`).

tl;dr;: it's good as it is now.

File: tools/Shared/Microsoft.TemplateEngine.CommandUtils/DotnetCommand.cs (hunk 4 above)
Comment (JanKrivanek) [reply tools/Shared/Microsoft.TemplateEngine.CommandUtils/DotnetCommand.cs:16]: On of the TODOs for the Templates Verification FW is addition of telemetry (tailored to usage of the authoring tooling) - that way (if we exclude that one as well in our integration tests) - we'd be able to get the idea of how much it's been used by other authors - and hence estimate the diff gap caused by excluding the dotnet CLI telemetry.

Comment (vlada-shubina) [reply tools/Shared/Microsoft.TemplateEngine.CommandUtils/DotnetCommand.cs]: After discussion we agreed to make telemetry output optional. @JanKrivanek already has a plan to implement bypassing env variables in test framework shortly.
